drivers/gles3/shaders/scene.glsl

```diff
@@ -1044,6 +1047,13 @@ void main() {
 #ifdef BASE_PASS
 	/////////////////////// LIGHTING //////////////////////////////
 
+#ifdef DEPTH_USED
+	vec3 ndc = vec3(screen_uv * 2.0 - 1.0, gl_FragDepth);
```

Comment:
I looked into optimizing this a little bit as we are throwing away half the result of the matrix multiplication. Doing a little investigation I found that the compiler does optimize away a few instructions because the results are thrown out, but it still loads the entire matrix in SGPRs which results in a few extra memory operations and an increased SGPR usage. 

A cheaper alternative is to just rely on z_near and z_far to linearize the depth as we do in the post processing effects. For example:
https://github.com/godotengine/godot/blob/d2a6a187f45345f00214c746924c57752a1f5260/servers/rendering/renderer_rd/shaders/effects/copy.glsl#L244-L256
I am a little skeptical that using this code will work, but it appears to work in the rest of the engine, so it would be a good idea to try. If it works it will save quite a few instructions and memory operations.

Reply:
I got the above logic working on Vulkan, it is indeed much simpler. The `z_near` and `z_far` values are available from `scene_data`.

[The NDC's Z component differs between OpenGL and Vulkan.](https://johannesugb.github.io/gpu-programming/why-do-opengl-proj-matrices-fail-in-vulkan/)

I guess that's why this same logic returns wrong linear depth in case of the GLES3 backend. Fixing it.

Reply:
[Found this post useful](https://stackoverflow.com/questions/11277501/how-to-recover-view-space-position-given-view-space-depth-value-and-ndc-xy/46118945#46118945)

According to this the GLES3 calculation should be:

`vertex.z = 2.0 * scene_data.z_near * scene_data.z_far / (scene_data.z_far + scene_data.z_near - (2.0 * gl_FragDepth - 1.0) * (scene_data.z_far - scene_data.z_near));`

Testing puts the carved out sphere closer to the camera, so the calculated view space Z is still wrong.

Any ideas?

Pushed the code, so you can take a look. Thanks.

Reply:
Since the interpretation of DEPTH differs between Vulkan and GLES3 the user fragment shader must use different DEPTH calculation, which is a problem. Right now it needs extra code to shuffle over the driver selection from ProjectSettings and pick the right calculation. 

It clearly shows that requiring a LINEAR_DEPTH instead of DEPTH would be both easier to use for game developers and working the same way over platforms, since the difference in DEPTH calculation would be shifted to the engine code.

Reply:
"Whether gl_FragCoord.z is linear or not depends on, the projection matrix. While for Orthographic Projection gl_FragCoord.z is linear, for Perspective Projection it is not linear." - [answer](https://stackoverflow.com/questions/7777913/how-to-render-depth-linearly-in-modern-opengl-with-gl-fragcoord-z-in-fragment-sh/45710371#45710371)

We cannot simplify it here without having to consider the current projection mode. 

Also, I don't know how it worked yesterday. I recompiled my code today and it failed to work, whatever I do. Likely I got the previous compiled shader from cache or whatever.

Anyway, reverted the optimization, back to the working code based on the projection matrix.

I cannot test the opengl3 driver, because it does not give me any shadows from the directional light. I don't know why.